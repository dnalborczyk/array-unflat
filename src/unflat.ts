function unflat<T>(value: readonly T[], size: 5): [T, T, T, T, T][]
function unflat<T>(value: readonly T[], size: 4): [T, T, T, T][]
function unflat<T>(value: readonly T[], size: 3): [T, T, T][]
function unflat<T>(value: readonly T[], size: 2): [T, T][]
function unflat<T>(value: readonly T[], size: 1): [T][]
function unflat<T>(value: readonly T[], size: number): T[][]
function unflat<T>(value: readonly T[]): [T, T][]

function unflat<T>(value: readonly T[], size: number = 2): T[][] {
  const unflattened: T[][] = []

  for (let i = 0; i < value.length; i += size) {
    const slice = value.slice(i, i + size)

    unflattened.push(slice)
  }

  return unflattened
}

export default unflat
