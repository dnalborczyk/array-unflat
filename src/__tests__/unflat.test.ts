import unflat from '../unflat'

interface Test {
  description: string
  expected: unknown[]
  param: unknown[]
  paramSize?: number
}

const tests: Test[] = [
  {
    description: '',
    expected: [],
    param: [],
  },

  {
    description: '',
    expected: [['a']],
    param: ['a'],
  },

  {
    description: '',
    expected: [['a', 1]],
    param: ['a', 1],
  },

  {
    description: '',
    expected: [['a', 1], ['c']],
    param: ['a', 1, 'c'],
  },

  {
    description: '',
    expected: [
      ['a', 1],
      [true, {}],
      [undefined, null],
    ],
    param: ['a', 1, true, {}, undefined, null],
  },

  {
    description: '',
    expected: [],
    param: [],
    paramSize: 1,
  },

  {
    description: '',
    expected: [],
    param: [],
    paramSize: 2,
  },
]

describe('unflat', () => {
  tests.forEach(({ description, expected, param, paramSize }) => {
    test(`should return ${description}`, () => {
      const result = unflat(param, paramSize)
      expect(result).toEqual(expected)
    })
  })
})
