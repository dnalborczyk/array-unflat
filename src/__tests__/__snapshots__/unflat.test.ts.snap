// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unflat exceptions should throw exception if size is not a number 1`] = `"unflat size parameter must be a number."`;

exports[`unflat exceptions should throw exception if size is null 1`] = `"unflat size parameter must be a number."`;

exports[`unflat exceptions should throw exception if value is not an array 1`] = `"unflat value parameter must be an array."`;
